### ISSUE TITLE
Toggle between most recent scripts/files

### ISSUE THREAD

**Issue by pat-s** [rstudio/rstudio#5804]
Similar to CTRL+Tab for browsers: Toggle between the current and previously active script. 

I often jump around between two files in a similar fashion one jumps between two browser tabs.
A toggle option for this would be awesome.

**Comment by dfalty** [comment]
Hi @pat-s, you can actually do this with ctrl+tab and ctrl+shift+tab to move between tabs in the source pane. Is this what you mean or am I missing something?
![image](https://user-images.githubusercontent.com/15237845/70001337-a0db0100-1522-11ea-95ee-59e7f81235d1.png)

**Comment by pat-s** [comment]
@dfalty Thanks. I am aware of both but it feels intuitively for me to use two shortcuts for it. Also I am used to the "toggle" shortcut CTRL+TAB in many other apps.

Basically, I just would like to have a "toggle" one to switch between the current and last tab like the already integrated toggle shortcuts.

**Comment by dfalty** [comment]
@pat-s Ah, I understand now, thanks for the explanation. If we added the ability for ctrl+tab and ctrl+shift+tab to "round-robin" around it would make this a better experience so you could go through all tabs with one shortcut. I thought we had a feature request for this somewhere but I was unable to find it. This would work in a scenario where you only have two tabs open but not if you have a bunch.

**Comment by pat-s** [comment]
Just to clarify: With "round-robin" you mean "go through all tabs and start at the beginning again"? 

I meant to have a toggle for "back and forth" between the current and most recent tab. Just like CTRL+Tab does in Chrome. 
Maybe you already meant this and I just didn't understand fully :)

**Comment by dfalty** [comment]
Yeah, I do mean "go through all tabs and start at the beginning again". It looks like this is actually the behavior Chrome has with ctrl+tab unless you have an extension that changes it. I can understand both use cases so I'll let others weigh-in on implementation. Thanks for the enhancement request!

**Comment by pat-s** [comment]
Oh yeah, my bad - I have indeed an extension mapping this custom behavior. But yeah, both would be great! Thanks for taking this on!

**Comment by jmcphers** [comment]
> Yeah, I do mean "go through all tabs and start at the beginning again". 

FWIW this has been implemented for a long time. In Tools -> Options -> General, check "Wrap around when navigating to previous/next tab".

**Comment by dfalty** [comment]
D'OH! Thanks @jmcphers

**Comment by SimonDedman** [comment]
Hi folks,
Just some clarification here since nothing has happened with this and it's easy to conclude from the last 2 comments that this is already available, which it isn't:
@pat-s and I (#7576 dupe) are asking for crtl+tab to switch to the last-open tab, regardless of their position in the top bar.
@dfalty asked about round-robin and pat and he discussed this and @jmcphers informed them this already exists which is nice & all BUT, this isn't the feature request pertaining to this issue.

Example for further clarity:
I'm editing script/tab 2 of 8 open tabs, which depends on tab 6, so I'm switching between them. From tab2, ctrl+tab would jump to the last open tab which is 6. From tab 6, ctrl+tab would jump to the last open tab which is tab2.

For extra credit, Firefox (& perhaps others) have added functionality whereby if you hold down the crtl key and keep hitting tab without releasing the ctrl key, you'll cycle through previous tabs, again regardless of the order they appear in the panel/toolbar. This seems to be limited to 6 which is probably some sensible tradeoff (since firefox has just shamed me into realising I have 81 tabs open).

**Comment by SimonDedman** [comment]
@jmcphers Jonathan any chance of a bump for this? Would be great to have and feels like a relatively easy thing to code... just seems to have fallen between the cracks of github issue triage... Cheers!

**Comment by MarcinKosinski** [comment]
+1

**Comment by N1h1l1sT** [comment]
+1

I mean, this is tremendously important for my sanity and productivity every time I use RStudio.

**Comment by GitHunter0** [comment]
> 
> 
> +1
> 
> I mean, this is tremendously important for my sanity and productivity every time I use RStudio.

I agree

**Comment by pat-s** [comment]
Unstale

**Comment by vadim-cherepanov** [comment]
The complete absence of this option makes it a real pain to work with two-three files at a time when one needs to jump back and forth between them, especially having some more project files open at the same time. The best you can do now is 1) move the tabs of these 2-3 files in one place (often disrupting the logic of the way they are placed first of all), and 2) each time you need to jump back to a previous file, you need to check whether its tab is to the left or right of the current one, and how many positions from it, and then press Ctrl+Tab or Ctrl+Shift+Tab correspondingly. And still you can miss the right key combination, e.g. simply because you are focused on your work rather than figuring out the relative position of some tabs and how it affects which keys you need to press in the IDE.

To make it more clear, consider the behavior of Alt(Command)+Tab in Linux, Mac or Windows -- the windows are sorted in the order of most recently used, and the shortcut switches windows in this order. It would be a real pain if to switch between two apps, say RStudio and the browser, one would constantly need to figure out where the other app is on the taskbar, "to the left" or "to the right", and then press Alt+Tab or Alt+Shift+Tab correspondingly, and then be completely lost after switching to a third app, say a chat...

Now, I understand that there are different scenarios. In some cases the predictability of the Ctrl[+Shift]+Tab behavior comes from the order of the tabs, so that one can just look at the tabs and without knowing what was open before know what to press and how many times to get to another file. However, often, when working with a small subset of files in a larger project, it helps to quickly switch back and forth between these files in the way similar to the OS. So, I think, while it may be wrong to simply create this behavior by default, there should at least be an option, or even better an alternative shortcut to switch tabs in the most recent order.

**Comment by GitHunter0** [comment]
@vadim-cherepanov , yes, I believe this issue is a no brainer, it is that kind of simple feature that makes a huge difference in the workflow.

**Comment by kevinushey** [comment]
For what it's worth, we have these back + forward buttons for navigating between source locations, which corresponds to edit history across documents:

![image](https://user-images.githubusercontent.com/1976582/167987483-638ba246-1c79-4588-9103-20f5fc1d9181.png)

We could consider extending this in some way to navigate to the last edit (source navigation) that placed focus into a different editor tab.

**Comment by GitHunter0** [comment]
Nice, my suggestion would be to use something like `Ctrl+Shift+F8` to toggle between the last locations/lines in the active file, and `Ctrl+Tab` to toggle between the last 2 active files.

**Comment by SimonDedman** [comment]
+1 for the crtl+tab option.

Kev I wonder if requiring edits might be a suboptimal route to take, e.g. when users are switching from one tab they're reading/copying but not actively editing, and another they're building. Wouldn't the edits requirement mean the first tab isn't 'active' in that sense?

**Comment by kevinushey** [comment]
> Wouldn't the edits requirement mean the first tab isn't 'active' in that sense?

Sorry, I misspoke -- it's really any general navigation; not just edits. So it should still work in that case (even though you'd have to click it a bunch of times to get to the last active tab).

Either way, I agree adding some commands for navigating between tabs in this way would be useful.

**Comment by kevinushey** [comment]
Adding to triage as I think the OP's request was initially understood (we want to navigate tabs based on the user navigation history, not just their visual position in the Source pane)

**Comment by joelsand** [comment]
Agreed, I would love for Ctrl+Tab to behave like it does in Visual Studio and how Alt+Tab does in Windows.

**Comment by mschuessler** [comment]
+1
I have the feature in any browser or edition I am using. I can't not understand why RStudio does not have it.

**Comment by GitHunter0** [comment]
> +1 I have the feature in any browser or edition I am using. I can't not understand why RStudio does not have it.

I don't get it either, this should be an easy feature to implement with considerable productivity gains.
I guess the "likes" is what is driving the priority and this issue unfortunately doesn't have many.

**Comment by SimonDedman** [comment]
I somewhat suspect the initial/ongoing confusion regarding what this request is, contributed to losing any early enthusiasm - it dragged on and got forgotten about.

**Comment by ronblum** [comment]
FWIW, at one point we targeted this fix for the last release, but we ended up putting it back in the backlog as part of the ruthlessness.

But—if anyone would like to code this feature, we're glad for y'all to contribute to RStudio! Whether as an add-in or in RStudio itself. For the latter, we act as the gatekeepers—we review submitted PRs, provide feedback as necessary, etc. RStudio is an open source project, and we both anticipate and welcome community contributions! 😃

**Comment by venpopov** [comment]
Bumping this up as I would love the feature to switch between two active tabs. When I work on a package, I might have a dozen files open, and for example switching back and forth between an .R file with a function and the corresponding test-*.R file. Having ctrl+tab work like alt+tab on windows will be wonderful.

**Comment by grohalex** [comment]
**+1**
Please develop this feature, it would 100% help with navigating R studio! Can't believe that every user isn't requesting this functionality of "opening the last active tab"

**Comment by SimonDedman** [comment]
Any chance of this being reinserted into the todo list at some point gents? Looks like it was added to milestones but doesn't seem to have been done and I think therefore has been pushed into Later. Hopefully it's not a ton of work - notwithstanding we know y'all have an increasingly insane backlog of stuff that users want!

Cheers chaps.

**Comment by SimonDedman** [comment]
@ronblum please please please please can someone put this on the todo list? See also this thread: https://github.com/rstudio/rstudio/issues/7576
